src/auth/permissions/index.ts

```diff
@@ -0,0 +1,173 @@
+import * as knex from 'knex';
+import { PermissionsInitialiser, GetPermissionIds, GetRoleId } from './permissions';
+
+const getPermissionIds: GetPermissionIds = ({ client, resource, permissionLevel, access }) =>
+  client
+    .select('permission_id')
+    .table('permission')
+    .where({
+      ['permission_entity']: resource,
+      ['permission_level']: permissionLevel,
+      ['access_type']: access,
+    });
+
+const getRoleId: GetRoleId = async({ client, role }) =>
+  client
+    .select('access_role_id')
+    .table('access_role')
+    .where({ ['access_role_name']: role })
+    .then((res) => res[0].access_role_id);
+
+const permissionsInitialiser: PermissionsInitialiser = (config) => {
+  const client = knex(config);
+  return {
+    grantExisting: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 1) {
+        const permissionId = permissionIds[0].permission_id;
+
+        const insertPermission = await client('access_role_permission')
+            .returning('*')
+            .insert(
+              client.select({
+                ['access_role_id']: roleId,
+                ['permission_id']: permissionId,
+              })
+                .whereNotExists(client('access_role_permission').where({
+                  ['access_role_id']: roleId,
+                  ['permission_id']: permissionId,
+                }))
+            );
+        if (insertPermission.length === 0) {
+          throw new Error('Permission entry is already associated to this role');
+        }
+        client.destroy();
+        return insertPermission;
+      } else {
+        client.destroy();
+        throw new Error('Permission entry does not exists, please use grantNew method');
+      }
+    },
+    grantNew: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 0) {
+        const createPermissionId = await client('permission')
+            .returning('permission_id')
+            .insert({
+              ['permission_entity']: resource,
+              ['permission_level']: permissionLevel,
+              ['access_type']: access,
+            });
+
+        const permissionId = createPermissionId[0];
+        const insertPermission = await client('access_role_permission')
```

Comment:
Maybe this and the above can be combined into a single query, using a common table expression (CTE), using [with](https://knexjs.org/#Builder-with):
```js
const insertPermission = client('permission').returning('*').insert({...})
client
  .with('new_permission', client.raw(insertPermission.sql, insertPermission.bindings))
  .insert({
    ['access_role_id']: roleId,
    ['permission_id']: select('permission_id').from('new_permission')
  })
```
Basically I'm trying to get this SQL:
```sql
WITH (
  INSERT INTO permission (...) VALUES (...) RETURNING permission_id
) AS new_permission_id
INSERT INTO access_role_permission (...) VALUES (..., new_permission_id) RETURNING *
```
The benefit is that the above is run as a single, all-or-nothing query without having to resort to a full-on transaction. As the code is currently written, there's a small chance we could be left with an inconsistent state in the DB if something goes wrong halfway through.

Reply:
If any operation consists of multiple DB queries, we should always think about data consistency, and whether we can combine the queries to be run as a single atomic block.

That usually means either sub-queries, CTEs or transactions.